docs/rfcs/20220413-aot-for-all-snode.md

```diff
@@ -0,0 +1,340 @@
+* By field ID: You can also use the field ID returned by `add_field()` to access a field. Here's an example:
+   ```py
+   builder = ti.SNodeTreeBuilder()
+   x_fid = builder.add_field(dtype=ti.f32, name='x')
+   # boilerplate to generate tree type and instantiate a tree ...
+
+   @ti.kernel
+   def foo(tr: tree_t):
+     x = ti.static(tr.get_field(x_fid))  # 1
+     for i in x:
+       x[i] = i * 2.0
+   ```
+
+   Note that this design requires that part of the kernel (1) being evaluated inside Python. It also pulls in the global variable `x_fid`, which kind of violates our initial goal. We could require that `x_fid` is passed into the kernel as an argument. But maybe it's fine just to view that as a trivial Python constant?
+
+## AoS vs SoA
+
+Two composite types require the switch between AoS vs SoA, `ti.Matrix` and `ti.Struct`.
+
+AoS is quite straightforward. One can just use the composite type as the `dtype` of the field. For example:
+
+```py
+builder = ti.SNodeTreeBuilder()
+
+x = builder.add_field(dtype=ti.vec3, name='x')  # ti.vec3 is a vector of 3 ti.f32's
+builder.dense(ti.i, 8).place(x)
+tree_t = x.build()
+```
+
+For SoA, things get a bit trickier. The current approach is to treat each compopnent of the composite type as a standalone scalar Taichi field. In the example below, we have to manually place the underlying 3 components of `x` separately.
+
+```py
+x = ti.Vector.field(3, ti.f32)
+for f in x._get_field_members():  # `x` consists three scalar f32 fields
+  ti.root.dense(ti.ij).place(f)
+```
+
+This introduces confusion at several spots:
+
+1. Type is not purely decided by `dtype`, but also by how the field is placed.
+2. It introduces the notion of "nested field", which Taichi doesn't currently have a good abstraction for. Because of this, it is quite complicated to apply certain kind of optimizations for a composite-typed field. For example, vectorized load/save consumes the same bandwidth as scalar ops on certain platforms. Without a good abstraction, the checking for whether a matrix field is AoS or SoA has to be spread across different passes in CHI IR.
+
+If we further think about the problem, SoA `x` is not really a field. Instead, it is a *grouped view* of three individual scalar fields. This view provides matrix operations, which won't make sense for each individual scalar field.
+
+In addition, because type is currently coupled with Taichi field definition, a Taichi field has to be implemented as individual fields in order to support the SoA scenario. Once we switch to the type builder pattern, we can control how the type is constructed first, and choose the field implementation later.
+
+
+```py
+builder = ti.SNodeTreeBuilder()
+vel = builder.add_field(dtype=ti.vec3, name='vel', layout=ti.Layout.SOA)
```

Comment:
In fact, the case you are describing is not `SOA`, but any layout other than `AOS`. I suggest rephrasing this section a bit. Using `ti.Layout.SOA` here is especially inaccurate.

Reply:
SG!

Reply:
I just removed this part. `add_field` should not be mixed with `add_field_view`.